icc-x-api/icc-crypto-x-api.ts

```diff
@@ -913,51 +978,72 @@ export class IccCryptoXApi {
     return lsItem !== null ? this.utils.base64toByteArray(lsItem) : null
   }
 
-  generateKeyForDelegate(ownerId: string, delegateId: string) {
-    return Promise.all([
-      this.getHealthcareParty(ownerId),
-      this.getHealthcareParty(delegateId)
-    ]).then(
-      ([owner, delegate]) =>
-        delegate.publicKey
-          ? this.AES.generateCryptoKey(true)
-              .then(AESKey => {
-                const ownerPubKey = utils.spkiToJwk(utils.hex2ua(owner.publicKey!))
-                const delegatePubKey = utils.spkiToJwk(utils.hex2ua(delegate.publicKey!))
-
-                return Promise.all([
-                  this.RSA.importKey("jwk", ownerPubKey, ["encrypt"]),
-                  this.RSA.importKey("jwk", delegatePubKey, ["encrypt"])
-                ]).then(([ownerImportedKey, delegateImportedKey]) =>
-                  Promise.all([
-                    this.RSA.encrypt(ownerImportedKey, this.utils.hex2ua(AESKey as string)),
-                    this.RSA.encrypt(delegateImportedKey, this.utils.hex2ua(AESKey as string))
-                  ])
+  generateKeyForDelegate(
+    ownerId: string,
+    delegateId: string
+  ): Promise<models.HealthcarePartyDto | models.PatientDto> {
+    const genProm = new Promise<
+      [null | "hcp" | "patient", models.HealthcarePartyDto | models.PatientDto]
+    >((resolve, reject) => {
+      //Those caches are invalidated first so that subsequents call to hcPartiesRequestsCache or hcPartyKeysRequestsCache take into account the newly generated keys
+      //And we do not have parallel processes for keys generations
+      // invalidate the hcp cache for the modified hcp
+      this.hcPartiesRequestsCache[ownerId] = {
+        entityType: null,
+        entity: genProm.then(hcpOrPat => {
```

Comment:
The promise constructor uses the promise itself, which does not exist yet. Even if the initialization were delayed, this simply cannot work (see below).